### PR TITLE
[SPARK-35439][SQL][FOLLOWUP] ExpressionContainmentOrdering should not sort unrelated expressions

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This is a followup of #32586. We introduced `ExpressionContainmentOrdering` to sort common expressions according to their parent-child relations. For unrelated expressions, previously the ordering returns -1 which is not correct and can possibly lead to transitivity issue.

### Why are the changes needed?

To fix the possible transitivity issue of `ExpressionContainmentOrdering`.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 1]
```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
...
+/**
+ * Orders `Expression` by parent/child relations. The child expression is smaller
+ * than parent expression. If there is child-parent relationships among the subexpressions,
+ * we want the child expressions come first than parent expressions, so we can replace
+ * child expressions in parent expressions with subexpression evaluation. Note that
+ * this is not for general expression ordering. For example, two irrelevant expressions
+ * will be considered as e1 < e2 and e2 < e1 by this ordering. But for the usage here,
+ * the order of irrelevant expressions does not matter.
+ */
+class ExpressionContainmentOrdering extends Ordering[Expression] {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:221]: Just curious, is there a reason of this move?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 2]
```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
...
   }
 }
+
+/**
+ * Orders `Expression` by parent/child relations. The child expression is smaller
+ * than parent expression. If there is child-parent relationships among the subexpressions,
+ * we want the child expressions come first than parent expressions, so we can replace
+ * child expressions in parent expressions with subexpression evaluation. Note that
+ * this is not for general expression ordering. For example, two irrelevant expressions
+ * will be considered as e1 < e2 and e2 < e1 by this ordering. But for the usage here,
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:218]: Shall we change this to `0` according to the new logic?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:221]: Oh, as it is a nested class, I cannot allocate it separately, but

```scala
val equivalence = new EquivalentExpressions
val exprOrdering = new equivalence.ExpressionContainmentOrdering
```

I can revert to nested class if you think it's unnecessary change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala]: Right, missing the doc. Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:221]: Never mind. New one also looks good~

**Comment by dongjoon-hyun** [comment]
Could you rebase to the master branch? The linter failure was fixed on the master branch.

**Comment by viirya** [comment]
Rebased. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 3]
```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
...
 }
+
+/**
+ * Orders `Expression` by parent/child relations. The child expression is smaller
+ * than parent expression. If there is child-parent relationships among the subexpressions,
+ * we want the child expressions come first than parent expressions, so we can replace
+ * child expressions in parent expressions with subexpression evaluation. Note that
+ * this is not for general expression ordering. For example, two irrelevant expressions
+ * will be considered as equal by this ordering. But for the usage here, the order of
+ * irrelevant expressions does not matter.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:219]: To be complete, could you add some description about the semantically-equal expressions?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:219]: Sure. Added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 4]
```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
...
+class ExpressionContainmentOrdering extends Ordering[Expression] {
+  override def compare(x: Expression, y: Expression): Int = {
+    if (x.find(_.semanticEquals(y)).isDefined) {
+      // `y` is child expression of `x`.
+      1
+    } else if (y.find(_.semanticEquals(x)).isDefined) {
+      // `x` is child expression of `y`.
+      -1
+    } else {
+      // Irrelevant expressions
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:230]: ditto. We should mention the semantically-equal expression here.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (only one minor comment, https://github.com/apache/spark/pull/32870/files#r649667440)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala]: added. thanks.

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thank you. Merged to master.

**Comment by Kimahriman** [comment]
I think this could theoretically still cause some issues because it doesn't follow the last rule for comparators:

>Finally, the implementor must ensure that compare(x, y)==0 implies that sgn(compare(x, z))==sgn(compare(y, z)) for all z.

A simple example I found that doesn't sort correctly:

```
val add1 = Add(Literal(1), Literal(2))
val add2 = Add(Literal(2), Literal(3))
val addParent = Add(add1, Literal(4))
val exprs = Seq(addParent, add2, add1)
assert(exprs.sorted(exprOrdering) === Seq(add2, add1, addParent))
```

The result remains `addParent, add2, add1`. I think because `compare(addParent, add2) == 0` and `compare(add2, add1) == 0`, essentially the list is already sorted. Whether in practice this only could lead to a suboptimal sort or could still cause a sorting exception like I saw previously I'm not sure.

**Comment by viirya** [comment]
I noticed that, but currently I have not better idea to sort the expressions better. For irrelevant expressions, seems no good rule to order them in deterministic way. Right now I just can make it meet transitivity contract so it can avoid the exception. As mentioned in its doc, this is not for general expression ordering but just for the specific usage. I think it seems to be rare to produce suboptimal sort. I'll think if there is better way to sort it.

**Comment by Kimahriman** [comment]
In my fork I just changed it to 

```
.sortBy(_.head.collect({ case e => e }).size)
```

so it basically just sorts by the number of expressions in the tree (not sure if there's an easier way to get that count than how I did it). Haven't done exhaustive testing on it but I feel like that makes sense to do. Not sure how one expression could contain another if it doesn't have more total expressions

**Pull request closed without merging**